ioSPI/fourier.py

```diff
@@ -0,0 +1,237 @@
+import numpy as np
+import pyfftw.interfaces.numpy_fft
+from numba import jit
+
+
+@jit
+def make_neg_pos_2d(arr2d_shape):
+    """Apply checkerboard pattern for 2D FFT.
+
+    Each pixel switches from positive to negative in checker board pattern.
+    Equivalent to fft shifting
+    https://dsp.stackexchange.com/questions/9039/centering-zero-frequency-for-discrete-fourier-transform
+
+    Parameters
+    ----------
+    arr2d_shape : tuple (n_particle,N_fourier_pixels,N_fourier_pixels)
+        Input shape tuple.
+
+    Returns
+    -------
+    neg_pos_2d : numpy.ndarray, shape (n_particle,N_fft_pix,N_fft_pix)
+        Checkerboard array.
+    """
+    # assert arr2d.ndim == 3  # extra axis
+    n_particle, R1, R2 = arr2d_shape
+    neg_pos_2d = np.ones(arr2d_shape)
+    for particle in range(n_particle):
+        for r1 in range(R1):
+            for r2 in range(R2):
+                if (r1 + r2) % 2:
+                    neg_pos_2d[particle, r1, r2] *= -1
+    return neg_pos_2d
+
+
+@jit
+def make_neg_pos_3d(arr3d_shape):
+    """Apply checkerboard pattern for 3D FFT.
+
+    Each pixel switches from positive to negative in checker board pattern.
+    Equivalent to fft shifting
+    https://dsp.stackexchange.com/questions/9039/centering-zero-frequency-for-discrete-fourier-transform
+
+    Parameters
+    ----------
+    arr3d_shape : tuple (N_fft_pix,N_fft_pix,N_fft_pix)
+        Input shape tuple.
+
+    Returns
+    -------
+    neg_pos_3d : numpy.ndarray, shape (N_fft_pix,N_fft_pix,N_fft_pix)
+        Checkerboard array.
+    """
+    R1, R2, R3 = arr3d_shape
+    neg_pos_3d = np.ones(arr3d_shape)
+    for r1 in range(R1):
+        for r2 in range(R2):
+            for r3 in range(R3):
+                if (r1 + r2 + r3) % 2:
+                    neg_pos_3d[r1, r2, r3] *= -1
+    return neg_pos_3d
+
+
+def fft3d(
+    arr3d,
+    mode,
+    neg_pos_3d=None,
+    numpy_fft=pyfftw.interfaces.numpy_fft,
+    only_real=False,
+):
+    """3D FFT
```

Comment:
Docstring one-liner should start with a verb at infinitiv: `Compute 3D FFT`?

Reply:
Please fix.

Reply:
Done